rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,631 @@
+                                op: match message_type {
+                                    4 => MQTTOperation::PUBACK(msgidonly),
+                                    5 => MQTTOperation::PUBREC(msgidonly),
+                                    6 => MQTTOperation::PUBREL(msgidonly),
+                                    7 => MQTTOperation::PUBCOMP(msgidonly),
+                                    _ => MQTTOperation::UNASSIGNED,
+                                },
+                            };
+                            Ok((&input[skiplen+len..], msg))
+                        }
+                        Err(e) => Err(e),
+                    }
+                }
+                8 => match parse_subscribe(rem, protocol_version) {
+                    Ok((_rem, subs)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::SUBSCRIBE(subs),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                9 => match parse_suback(rem, protocol_version) {
+                    Ok((_rem, suback)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::SUBACK(suback),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                10 => match parse_unsubscribe(rem, protocol_version) {
+                    Ok((_rem, unsub)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::UNSUBSCRIBE(unsub),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                11 => match parse_unsuback(rem, protocol_version) {
+                    Ok((_rem, unsuback)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::UNSUBACK(unsuback),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                12..=13 => {
+                    let msg = MQTTMessage {
+                        header: header,
+                        op: match message_type {
+                            12 => MQTTOperation::PINGREQ,
+                            13 => MQTTOperation::PINGRESP,
+                            _ => MQTTOperation::UNASSIGNED,
+                        },
+                    };
+                    if len > rem.len() {
+                        return Err(Err::Incomplete(Needed::Size(len - rem.len())));
+                    }
```

Comment:
Do we need this check for Incomplete ?

Reply:
I don't think so. I think it was left from earlier code.